File: 0150_draw_speed_comparison/canvas/canvas.clj
(ns canvas)

(gen-class
  :name "canvas"
  :main true
  :extends javafx.application.Application)

(import
  '(javafx.application   Application)
  '(javafx.event         EventHandler)
  '(javafx.scene         Group Node Scene)
  '(javafx.scene.input   KeyCode KeyEvent)
  '(javafx.scene.layout  BorderPane Pane)
  '(javafx.scene.paint   Color)
  '(javafx.scene.shape   Rectangle Polygon Circle
                         Path PathElement MoveTo ArcTo ClosePath
                         LineTo)
  '(javafx.stage         Stage))

(def origin (ref [0 0]))
(def grid 8.0)

(defn make-square [[x y]]
  (let [square (Rectangle. x y grid grid)]
    (doto square
      (.setStroke Color/BLACK)
      (.setFill Color/TRANSPARENT))
    [square
     (fn [[x y]]
       (.setX square x)
       (.setY square y)
       )]))

(defn make-circle [[x y]]
  (let [circle (Circle. (+ x (* 0.5 grid)) (+ y (* 0.5 grid)) (* 0.5 grid))]
    (doto circle
      (.setStroke Color/BLACK)
      (.setFill Color/TRANSPARENT))
    [circle
     (fn [[x y]]
       (.setCenterX circle (+ x (* 0.5 grid)))
       (.setCenterY circle (+ y (* 0.5 grid)))
       )]))

(defn make-triangle [[x y]]
  (let [triangle (Polygon. (double-array
                  [x y (+ x grid) (+ y (* 0.5 grid)) x (+ y grid)]
                  ))]
    (doto triangle
      (.setStroke Color/BLACK)
      (.setFill Color/TRANSPARENT))
    [triangle
     (fn [[x y]]
     ;  (.. triangle getPoints
     ;      (setAll (double-array
     ;               [x y (+ x grid) (+ y (* 0.5 grid)) x (+ y grid)]
     ;               ))))]))
       )]))

(defn make-and [[x y]]
  (let [move-to (MoveTo. x y)
        line-to1 (LineTo. (+ x (* 0.5 grid)) y)
        arc-to (ArcTo. (* 0.5 grid)       ; radiusX
                       (* 0.5 grid)       ; radiusY
                       0.0                ; AxisRotation
                       (+ x (* 0.5 grid)) ; x
                       (+ y grid)         ; y
                       false              ; largeArcFlag
                       true ; sweepFlag (true -> counter clockwise)
                       )
        line-to2 (LineTo. x (+ y grid))
        path (Path. (into-array PathElement
              [move-to line-to1 arc-to line-to2 (ClosePath.)]
              ))]
    (doto path
      (.setStroke Color/BLACK)
      (.setFill Color/TRANSPARENT))
    [path
     (fn [[x y]]
       (doto move-to (.setX x) (.setY y))
       (doto line-to1 (.setX (+ x (* 0.5 grid))) (.setY y))
       (doto arc-to (.setX (+ x (* 0.5 grid))) (.setY (+ y grid)))
       (doto line-to2 (.setX x) (.setY (+ y grid)))
       )]))

(defn change-position [ll [xorg yorg]]
  (doseq [[f [x y]] (mapcat (fn [l y]
                              (map (fn [f x] [f [x y]])
                              l
                              (iterate #(+ % grid) xorg)))
                            (partition 100 ll)
                            (iterate #(+ % grid) yorg))]
    (f [x y])))

(defn make-objects [[xorg yorg]]
  (mapcat (fn [y-in-grid]
            (let [y (+ (* y-in-grid grid) yorg)]
              (map (fn [x fobj] (fobj [x y]))
                   (take 100 (iterate #(+ % grid) xorg))
                   (drop (mod y-in-grid 4)
                         (cycle [make-square make-circle
                                 make-triangle make-and])))))
          (range 100)))

;--------------------------------------------------
; schematic pane
;--------------------------------------------------

(def key-commands
  {KeyCode/RIGHT (fn [] (alter origin update-in [0] #(+ grid %)))
   KeyCode/LEFT  (fn [] (alter origin update-in [0] #(- grid %)))
   KeyCode/UP    (fn [] (alter origin update-in [1] #(- grid %)))
   KeyCode/DOWN  (fn [] (alter origin update-in [1] #(+ grid %)))})

(defn pane-schem-key [f-set-to-parent ll]
  (proxy [EventHandler] []
    (handle [keyEvent]
      (let [f (key-commands (.getCode keyEvent))]
        (when f
          (dosync (f))
          (change-position ll @origin)
          (.consume keyEvent)
          )))))

(defn pane-schem [f-set-to-parent]
  (let [objects (make-objects @origin)
        pane (Pane.)]
    (.setOnKeyPressed pane
                      (pane-schem-key f-set-to-parent
                                      (map #(% 1) objects)))
    (.setFocusTraversable pane true)
    (.setAll (.getChildren pane)
             (into-array Node (map #(% 0) objects)))
    (f-set-to-parent pane)
    pane))

;--------------------------------------------------
; JavaFX main routine
;--------------------------------------------------
(defn -start [self stage]
  (let [topgroup (BorderPane.)
        pane (pane-schem #(.setCenter topgroup %))]
    (doto stage
      (.setWidth 1024) (.setHeight 768)
      (.setScene (Scene. topgroup))
      (.setTitle "Draws a lot of objects")
      (.show))
    (.requestFocus pane)))

(defn -main [& args]
  (Application/launch (Class/forName "canvas")
                      (into-array String [])))

File: 0150_draw_speed_comparison/canvas/canvas.java
import javafx.application.Application;
import javafx.event.EventHandler;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.stage.Stage;

public class canvas extends Application {
  double origin_x, origin_y, grid;
  Canvas canvas;
  GraphicsContext gc;

  void draw_square(GraphicsContext gc, double x, double y) {
    gc.strokeRect(x, y, grid, grid);
  }  

  void draw_circle(GraphicsContext gc, double x, double y) {
    gc.strokeOval(x, y, grid, grid);
  }

  void draw_triangle(GraphicsContext gc, double x, double y) {
    double[] xs = new double[3];
    double[] ys = new double[3];
    xs[0] = x; xs[1] = x + grid;     xs[2] = x;
    ys[0] = y; ys[1] = y + 0.5*grid; ys[2] = y + grid;
    gc.strokePolygon(xs, ys, 3);
  }

  void draw_and(GraphicsContext gc, double x, double y) {
    gc.beginPath();
    gc.moveTo(x, y);
    gc.lineTo(x + 0.5*grid, y);
    gc.arcTo( x + grid
            , y + 0.5*grid
            , x + 0.5*grid
            , y + grid
            , 0.5*grid );
    gc.lineTo(x, y + grid);
    gc.closePath();
    gc.stroke();
  }

  void draw_objects(GraphicsContext gc, double xorg, double yorg) {
    double x, y;
    y = yorg;
    for (int i = 0; i < 100; i++, y += grid) {
      x = xorg;
      for (int j = 0; j < 100; j++, x += grid) {
        int modulo = (j + i) % 4;
        switch (modulo) {
          case 0: draw_square(gc, x, y); break;
          case 1: draw_circle(gc, x, y); break;
          case 2: draw_triangle(gc, x, y); break;
          case 3: draw_and(gc, x, y); break;
        }
      }
    }
  }

  void installEventHandler(Canvas canvas) {
    EventHandler<KeyEvent> keyEventHandler =
    new EventHandler<KeyEvent>() {
      public void handle(KeyEvent keyEvent) {
        if      (keyEvent.getCode() == KeyCode.RIGHT) { origin_x += grid; }
        else if (keyEvent.getCode() == KeyCode.LEFT ) { origin_x -= grid; }
        else if (keyEvent.getCode() == KeyCode.UP   ) { origin_y -= grid; }
        else if (keyEvent.getCode() == KeyCode.DOWN ) { origin_y += grid; }
        //switch (keyEvent.getCode()) {
        //  case KeyCode.RIGHT: origin_x += grid; break;
        //  case KeyCode.LEFT : origin_x -= grid; break;
        //  case KeyCode.UP   : origin_y -= grid; break;
        //  case KeyCode.DOWN : origin_y += grid; break;
        //}
        gc.clearRect(0.0, 0.0, canvas.getWidth(), canvas.getHeight());
        draw_objects(gc, origin_x, origin_y);
        keyEvent.consume();
      }
    };

    canvas.setOnKeyPressed(keyEventHandler);
  }

  public void start(Stage stage) {
    origin_x = origin_y = 0.0; grid = 8.0;
    canvas = new Canvas(1024.0, 768.0);
    gc = canvas.getGraphicsContext2D();
    BorderPane topgroup = new BorderPane();
    gc.setStroke(Color.BLACK);
    gc.setFill(Color.TRANSPARENT);
    installEventHandler(canvas);
    canvas.setFocusTraversable(true);
    draw_objects(gc, origin_x, origin_y);

    topgroup.setCenter(canvas);
    stage.setWidth(1024);
    stage.setHeight(768);
    stage.setScene(new Scene(topgroup));
    stage.setTitle("Draws a lot of objects");
    stage.show();

    canvas.requestFocus();
  }

  public static void main(String[] args) {
    launch(args);
  }
}
